numpy/core/src/multiarray/descriptor.c

```diff
@@ -1486,84 +1461,86 @@ PyArray_DescrConverter(PyObject *obj, PyArray_Descr **at)
-        if (*at == NULL){
-            if (!PyErr_Occurred()) {
-                _report_generic_error();
-            }
-            return NPY_FAIL;
-        }
-        return NPY_SUCCEED;
+        return _convert_from_tuple(obj, 0);
     }
     else if (PyList_Check(obj)) {
         /* or a list */
-        *at = _convert_from_array_descr(obj,0);
-        if (*at == NULL) {
-            if (!PyErr_Occurred()) {
-                _report_generic_error();
-            }
-            return NPY_FAIL;
-        }
-        return NPY_SUCCEED;
+        return _convert_from_array_descr(obj, 0);
     }
     else if (PyDict_Check(obj) || PyDictProxy_Check(obj)) {
         /* or a dictionary */
-        *at = _convert_from_dict(obj,0);
-        if (*at == NULL) {
-            if (!PyErr_Occurred()) {
-                _report_generic_error();
-            }
-            return NPY_FAIL;
-        }
-        return NPY_SUCCEED;
+        return _convert_from_dict(obj, 0);
     }
     else if (PyArray_Check(obj)) {
         _report_generic_error();
-        return NPY_FAIL;
+        return NULL;
     }
     else {
-        if (_arraydescr_from_dtype_attr(obj, at)) {
+        PyArray_Descr *ret;
+        if (_arraydescr_from_dtype_attr(obj, &ret)) {
             /*
              * Using dtype attribute, *at may be NULL if a
              * RecursionError occurred.
              */
-            if (*at == NULL) {
-                return NPY_FAIL;
-            }
-            return NPY_SUCCEED;
+            return ret;
         }
         /*
          * Note: this comes after _arraydescr_from_dtype_attr because the ctypes
          * type might override the dtype if numpy does not otherwise
          * support it.
          */
         if (npy_ctypes_check(Py_TYPE(obj))) {
-            *at = _arraydescr_from_ctypes_type(Py_TYPE(obj));
-            return *at ? NPY_SUCCEED : NPY_FAIL;
+            return _arraydescr_from_ctypes_type(Py_TYPE(obj));
         }
         _report_generic_error();
-        return NPY_FAIL;
+        return NULL;
     }
 }
 
-static int
-_convert_from_bytes(PyObject *obj, PyArray_Descr **at)
+
+/*NUMPY_API
+ * Get typenum from an object -- None goes to NPY_DEFAULT_TYPE
+ * This function takes a Python object representing a type and converts it
+ * to a the correct PyArray_Descr * structure to describe the type.
+ *
+ * Many objects can be used to represent a data-type which in NumPy is
+ * quite a flexible concept.
+ *
+ * This is the central code that converts Python objects to
+ * Type-descriptor objects that are used throughout numpy.
+ *
+ * Returns a new reference in *at, but the returned should not be
+ * modified as it may be one of the canonical immutable objects or
+ * a reference to the input obj.
+ */
+NPY_NO_EXPORT int
+PyArray_DescrConverter(PyObject *obj, PyArray_Descr **at)
+{
+    *at = _convert_from_any(obj);
+    if (*at == NULL && !PyErr_Occurred()) {
+        _report_generic_error();
+    }
```

Comment:
unfortunately this really is reachable at the moment - a future patch will attempt to track down the cases when this happens, and emit the error in a more appropriate place.